### PR TITLE
Merkleprove opcode (Moved changes into development fork)

### PR DESCRIPTION
**Pull request by ansermino**
Moved this into a fork from development and fixed some important naming issues from previous attempt.

This is the code I have implemented in another fork of Geth. **I have not tested in this fork so please do so.**

1. Setup test directory:
- Clone https://github.com/shyftnetwork/shyft_tests into go-empyrean/test as testdata/ (technically its supposed to be a submodule at go-empyrean/tests/testdata but we want our test repo, will update submodule link later).

2. Run go test tests/

### CODE REVIEW

**Comment by ansermino** [comment]
@GregTheGreek @dutterbutter @noot waiting on approvals...

File: core/vm/instructions.go
Comment (GregTheGreek) [inline core/vm/instructions.go:939]: Any chance we can get a test for this please? we are blocking PR's unless a proper test is written @dbkbali confirm?

**Review by GregTheGreek — DISMISSED** [review]

File: core/vm/instructions.go
Comment (ansermino) [reply core/vm/instructions.go:939]: instructions_test.go only benchmarks and tests some future opcodes (LHS etc.). There is also gas_table_test.go which seems to include my addition automatically.

**Comment by GregTheGreek** [comment]
@dbkbali Are we fine to merge to development?

**Comment by dutterbutter** [comment]
@ansermino and I discovered last week that some of the tests fail for two reasons
1) We are not up to date with go-ethereum master which some of the test logic relies on
2) Our postgres instance interferes 

Not to sure what the best approach is when it comes to updating to reflect go-ethereum master

File: core/vm/instructions.go
Comment (dbkbali) [reply core/vm/instructions.go:939]: I was able to get the original contract tests to run by updating the submodule. @ansermino would it not make sense to fork the original submodule and add your shyft tests ?-We could then  update the submodule link to point to your fork. This would allow us to track changes relative to the original geth tests and determine incrementally whether your code is creating additional failures prior to merging into development.

**Comment by dbkbali** [comment]
See my comments above - I think a better way to do this would be to fork the original geth submodule tests branch - then from the forked submodule branch add the shyft tests - and update the submodule link to point to the forked repo.

**Comment by priom** [comment]
@dbkbali @GregTheGreek what's the update on this PR?

**Comment by ansermino** [comment]
@priom Waiting on testdbRefactor branch to be completed and merged into development to ensure I'm not introducing any new failures for the tests

**Comment by GregTheGreek** [comment]
Test db refactor has been finished. Are we experiencing the gas issue still?

**Comment by ansermino** [comment]
Gas fix is WIP

File: core/vm/instructions.go
Comment (noot) [inline core/vm/instructions.go]: could combine this with previous line

File: core/vm/instructions.go
Comment (noot) [inline core/vm/instructions.go]: instead of big.NewInt(32).Int64() would int64(32) work?

**Comment by ansermino** [comment]
@dbkbali @dutterbutter Please review and merge

**Comment by dutterbutter** [comment]
The way I checked to see if the tests are passing is as follows:
- switched to branch `david/merkleprove-dev`
- `cd tests`
- git cloned the above mentioned repo and renamed to testdata
- inside tests dir `go test ./...`
- in root `go test ./...`

All tests are passing on my machine, however, please use the linter on the above changed files. The linter commands that are run by the CI are the following:

**You will probably need to install gometalinter, see below for link**

`gometalinter ./... --vendor --tests --deadline=2m --disable-all --enable=goimports --enable=varcheck --enable=vet --enable=gofmt --enable=misspell --exclude=core/genesis.go --exclude=eth/handler_test.go --exclude=eth/downloader/api.go --enable=goconst --min-occurrences=6`

`gometalinter --vendor --tests --deadline=10m --disable-all --exclude=ethdb/shyft_database.go --enable=gosimple ./...`

`gometalinter --vendor --tests --deadline=10m --disable-all --enable=unconvert ./...`

To fix goimports, gofmt linting errors:
`goimports -w <filename>` // Must be in the appropriate directory,

For more details visit the lint repo [here](https://github.com/alecthomas/gometalinter)

Once linting changes have been made this PR should be good for merge.

**Review by dutterbutter — APPROVED** [review]

**Review by dbkbali — APPROVED** [review]
@dutterbutter @ansermino @GregTheGreek, All tests are passing however I did need to set -parallel to 1  on my machine - the reason for this is that we are still getting race conditions with the SQL driver when running on multiple cores (to eliminate the need for limiting parallel cores - at some point we probably need to refactor the pgdb tests to use https://github.com/ory/dockertest).

Thanks @ansermino for updating the submodule link, as a result there is no need to clone the evm tests in the tests folder instead you just need to:

```
cd tests
git submodule update --init --recursive
```
and that will bring in all the submodule tests into the tests/testdata folder

I had no need to run or install gometalinter - and if there is a need then we will probably have issues with travis.

I assume currently that Travis is not as part of its tasks updating the submodule and running these tests.

Moving forward it probably makes sense to update our travis config to do this so that we can be sure that merges are not causing our EVM to fail - appreciate your thoughts on this @GregTheGreek @ansermino @dutterbutter .. ie. Should we update the travis later or prior to this merge.

Otherwise I have no problems approving this request - as long as we understand that there is the possibility that travis will not currently be running the contract tests

**Comment by ansermino** [comment]
Thanks @dbkbali. I may note I had some issues figuring out how to lint, perhaps this is something we could add to the PR template or README? 

To clarify travis is running the submodule tests.

I would strongly advocate for trying some solutions for these travis fails, otherwise it is gonna be quite the headache down the line.

**Comment by tim-br** [comment]
I got tests to pass, and the changes look good.

**Review by tim-br — APPROVED** [review]

**Comment by GregTheGreek** [comment]
Lets go ahead and get this through. Looks like very little code changes (lines & files) so this is definitely a candidate to squash the changes. 

@ansermino we'll go over this quickly in person, but we'll squash your commits via the merge button (to preserve state on the PR itself)

**Pull request merged**
